### PR TITLE
Fix crash when cursor ends up at invalid position

### PR DESCRIPTION
**Pull request by dnfield**
Factored out a couple sanitization methods into FML so that Desktop can use them and we can unit test them more easily.

See https://github.com/flutter/flutter/pull/31468 and the linked issues from that for an attempt to fix this on the framework side.

Root cause:

- On iOS, it's possible to get the cursor at any byte position in the strong once you're at the second line of a multi-line text field (https://github.com/flutter/flutter/issues/24802)
- In this case, typing will end up inserting bytes into the middle of what is meant to be a single unicode sequence, creating an invalid UTF string.
- We were handling this property for incrementing or decrementing the position, but not for setting the selection.  The meat of this patch is adding the sanitiztion to set selection.
- When we go to encode or decode this to JSON, we hit an assert (because the JSON codec on iOS is not tolerant of invalid UTF), and the app crashes.

This fixes the input handling on iOS so we don't crash, and makes https://github.com/flutter/flutter/issues/24802 not be a crasher.  

Fixes https://github.com/flutter/flutter/issues/31525

### CODE REVIEW

**Comment by dnfield** [comment]
To be clear: we still need to fix the cursor positioning logic in Dart side, but this makes it slightly less bad.

File: fml/platform/darwin/string_range_sanitization.h [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+// found in the LICENSE file.
+
+#ifndef FLUTTER_FML_STRING_RANGE_SANITIZATION_H_
+#define FLUTTER_FML_STRING_RANGE_SANITIZATION_H_
+
+#include <Foundation/Foundation.h>
+
+namespace fml {
+
+// Helper to get the correct boundary of a character position in an NSString
```
Comment (cbracken) [inline fml/platform/darwin/string_range_sanitization.h]: Nit: here and below `|NSString|` etc. for identifiers.

File: fml/platform/darwin/string_range_sanitization.h [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+// given a byte index.
+NSRange rangeForCharacterAtIndex(NSString* text, NSUInteger index);
```
Comment (chinmaygarde) [inline fml/platform/darwin/string_range_sanitization.h]: Here and elsewhere, please use CapitalCase.

File: fml/platform/darwin/string_range_sanitization.h [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+#ifndef FLUTTER_FML_STRING_RANGE_SANITIZATION_H_
+#define FLUTTER_FML_STRING_RANGE_SANITIZATION_H_
+
+#include <Foundation/Foundation.h>
+
+namespace fml {
+
+// Helper to get the correct boundary of a character position in an NSString
+// given a byte index.
```
Comment (cbracken) [inline fml/platform/darwin/string_range_sanitization.h]: Consider rephrasing as `Returns a range encompassing the grapheme cluster containing in which |index| is located.` 

Generally, starting with a third-person verb tends to encourage comments that focus on what the code does. I'd also avoid the use of the word 'character' as it's pretty ambiguous. Similar below.

File: shell/platform/darwin/ios/framework/Source/flutter_text_input_plugin_unittest.mm [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+#include "flutter/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm"
+#include "gtest/gtest.h"
+
+TEST(FlutterTextInputView, SanitizesSelectedTextRange) {
+  FlutterTextInputView* view = [[FlutterTextInputView alloc] init];
+  [view insertText:@"😠"];
+  [view setSelectedTextRange:[FlutterTextRange rangeWithNSRange:NSMakeRange(1, 0)]];
+  FlutterTextRange* selectedRange = (FlutterTextRange*)[view selectedTextRange];
+  EXPECT_EQ(selectedRange.range.location, 2);
+}
```
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/flutter_text_input_plugin_unittest.mm]: Add tests for bogus indices.

File: fml/platform/darwin/string_range_sanitization.h (hunk 3 above)
Comment (dnfield) [reply fml/platform/darwin/string_range_sanitization.h]: The underlying impelementation throws an exception (`NSInvalidArgumentException`).  Is that fair for us to just do?

File: fml/platform/darwin/string_range_sanitization.mm [hunk 5]
```diff
@@ -0,0 +1,29 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/fml/platform/darwin/string_range_sanitization.h"
+
+namespace fml {
+
+NSRange RangeForCharacterAtIndex(NSString* text, NSUInteger index) {
+  if (text == nil || index > text.length) {
```
Comment (cbracken) [inline fml/platform/darwin/string_range_sanitization.mm]: I'd be tempted to use `index >= text.length` for consistency with Apple's APIs. It also feels a bit more intuitive: the underlying string has no character at `index == length`. 

That said, text positions _do_ allow for end-of-string positions. If every call to this is expected to have to defend against end-of-string positions, then it probably makes sense as-is, but I'd definitely document the range of valid indices in the header.

Comment (dnfield) [reply fml/platform/darwin/string_range_sanitization.mm]: I'll go with >=, good catch.

**Pull request merged**